apps/emqx/src/emqx_schema.erl

```diff
@@ -1882,6 +1902,25 @@ compile_variform(Expression, _Opts) ->
             throw(#{expression => Expression, reason => Reason})
     end.
 
+compile_topic_rule_match(Expression, Opts) ->
+    %% NOTE
+    %% Allow severely constrained variform expression as syntax for now, for
+    %% simplicity and predictability.
+    case compile_variform(Expression, Opts) of
+        Compiled = #{form := {call, TopicFn, [{var, "topic"}, {str, ArgStr}]}} when
```

Comment:
if we are only compiling the expression, but not actually rendering the result, there is perhaps no need to force the usage of var `topic`, after all, there is only one possible input which is the topic used in SUBSCRIBE request.

Reply:
The goal here is to keep it forward-compatible with variform (or some future superset of it) while having limited subset as a first iteration, and having implicit arguments might compromise it in my opinion. To be clear, this is an UX concern: I think it's better to (eventually) have a single expression language in such contexts rather than several slightly different ones.

Moreover, this struct may potentially be used in other contexts besides SUBSCRIBE requests (most of contexts that come to mind admittedly also have just one well-defined `topic`, but still).

Reply:
> I think it's better to (eventually) have a single expression language in such contexts rather than several slightly different ones.

What is variform? This is the only description of a programming language with this name I found on the internet: https://docs.emqx.com/en/emqx/latest/configuration/configuration.html#variform-expressions 

There's no BNF, no denotational semantics, just a couple of examples. It's not well known even among EMQX developers :wink: I suggest we postpone deep integration of this single expression language until it's well documented.

Reply:
> The primary goal is forward compatibility: if we need to introduce more bindings in the evaluation context (e.g. listener or clientid) in the future, and more predicates, then having implicit arguments is going to bite.

What will happen **currently** if I try to use these additional predicates? Crash? Rule will silenty fail to match?

Reply:
Currently there's no additional predicates, so it's an error that will cause configuration to be rejected as invalid.